docs/internals/writing-a-provider.rst

```diff
@@ -0,0 +1,134 @@
+===========================
+Writing an EF Core Provider
+===========================
+
+.. contents:: `In this article`
+  :local:
+
+EF Core is designed to be extensible. It provides general purpose building
+blocks that are intended for use in multiple providers. The purpose of this
+article is to provide basic
+guidance on creating a new provider that is compatible with EF Core.
+
+.. tip::
+  `EF Core source code is open-source <https://github.com/aspnet/EntityFramework>`_.
+  The best source of information is the code itself.
+
+.. tip::
+  This article shows snippets from an empty EF provider. You can view the `full
+  stubbed-out provider
+  <https://github.com/aspnet/EntityFramework.Docs/tree/master/docs/internals/sample>`_
+  on GitHub.
+
+.. _entry-point:
+
+DbContext Initialization
+------------------------
+
+A user's interaction with EF begins with the ``DbContext`` constructor. Before
+the context is available for use, it initializes **options** and **services**. We will example both of
+these to understand what they represent and how EF configures itself to use
+different providers.
+
+Options
+^^^^^^^
+
+``Microsoft.EntityFrameworkCore.Infrastructure.DbContextOptions`` is the API surface for **users** to configure ``DbContext``. Provider
+writers are responsible for creating API to configure options and to make
+services responsive to these options. For example, most providers require a
+connection string. These options are
+typically created using ``DbContextOptionsBuilder``.
+
+Services
+^^^^^^^^
+
```

Comment:
Ah, that is in the Understanding Services page... perhaps under the "Add Method" section you could call out that in some scenarios users will call the AddX method directly, and in other cases it will be called by EF.